### PR TITLE
glibc: Fix compiling with the brewed gcc for Linuxbrew

### PR DESCRIPTION
**Pull request by sjackman**
This PR permits upgrading `glibc`. To test it, add `revision 1` to the `glibc` formula and try `brew upgrade glibc`. I've also fixed installing `glibc` using `--env=super`.

### CODE REVIEW

**Comment by maxim-belkin** [comment]
This is what I get in a clean Docker container when I follow the steps @jonchang outlines above:
https://gist.github.com/1db4441e2646d4941d4a1a4d4048e1ae

**Comment by sjackman** [comment]
It's helpful to quote the error if you can, Maxim.
```
checking for gawk... no
configure: error: 
*** These critical programs are missing or too old: gawk
```
https://gist.github.com/anonymous/1db4441e2646d4941d4a1a4d4048e1ae#file-01-configure-L54

I encountered that problem as well. I'm hesitant to add a `default_formula "gawk"` to `glibc`, because that would mean that `gawk` and its dependencies would need to be built portable. One possibility of `glibc` allows it is to use brewed `mawk` rather than `gawk` and built a portable bottle for it. In any case, we should add a `GawkRequirement` to `glibc` that fails if `gawk` is not available. I'll do that.

**Comment by sjackman** [comment]
I've added `GawkRequirement` to this PR.
@maxim-belkin Could you please test it again?

File: Formula/glibc.rb [hunk 1]
```diff
@@ -1,3 +1,20 @@
+class GawkRequirement < Requirement
+  fatal true
+
+  satisfy(build_env: false) do
+    which "gawk"
```
Comment (maxim-belkin) [inline Formula/glibc.rb]: Well, we need a modern version of `awk`, not just `(g)awk`. 
In a Linuxbrew docker container:
```
$ /usr/bin/awk -W v
mawk 1.3.3 Nov 1996, Copyright (C) Michael D. Brennan

compiled limits:
max NF             32767
sprintf buffer      2040
```

On a Mac:
```
$ awk --version
awk version 20070501
```

**Comment by maxim-belkin** [comment]
Hmm, there is no `gawk` in my Docker container but requirement was not triggered:
```
checking for sed... sed
checking version of sed... 4.2.2, ok
checking for gawk... no
checking for nm... nm
configure: error: 
*** These critical programs are missing or too old: gawk
*** Check the INSTALL file for required versions.
```

**Comment by maxim-belkin** [comment]
would it make sense to factor out gawk requirement into a separate file in `Library/Homebrew/requirements`?

**Comment by sjackman** [comment]
So far this is the only occurrence of `GawkRequirement`. If it occurs in more than one formula (I'd say more than two), we can factor it out. `glibc` is also a special case, because we don't want it to have a `default_formula`, whereas for other formulae we would want a `default_formula`.

**Comment by sjackman** [comment]
> Hmm, there is no gawk in my Docker container but requirement was not triggered:

Woops. That was a bug. Should be fixed now. Thanks for catching it.

**Comment by sjackman** [comment]
Merged!

**Comment by maxim-belkin** [comment]
```
$ brew upgrade glibc
==> Upgrading 1 outdated package, with result:
glibc 2.19_1
glibc: gawk is required to build glibc.
Install gawk with your host package manager if you have sudo access.
  sudo apt-get install gawk
  sudo yum install gawk
Error: An unsatisfied requirement failed this build.
```

**Comment by jonchang** [comment]
I think the `glibc` in linuxbrew is still 2.19 (without a revision)

**Comment by maxim-belkin** [comment]
I omitted the step where I add `revision 1` to the formula

**Comment by rwhogg** [comment]
Sorry for being late chiming in here. One other issue on Ubuntu 17.04 (admittedly not a system where brewed `glibc` is all that helpful): The system GCC 6.3 is used by default and seems to fail, even if brewed GCC 5.3 is first in PATH:

```
/home/linuxbrew/.linuxbrew/opt/binutils/bin/ld: /tmp/glibc-20170810-4202-d41b88/glibc-2.19/build/csu/crt1.o: relocation R_X86_64_32S against symbol `__libc_csu_fini' can not be used when making a shared object; recompile with -fPIC
/home/linuxbrew/.linuxbrew/opt/binutils/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/6/crtbegin.o: relocation R_X86_64_32 against hidden symbol `__TMC_END__' can not be used when making a shared object
/home/linuxbrew/.linuxbrew/opt/binutils/bin/ld: final link failed: Nonrepresentable section on output
collect2: error: ld returned 1 exit status
../Rules:140: recipe for target '/tmp/glibc-20170810-4202-d41b88/glibc-2.19/build/iconv/iconv_prog' failed
make[2]: *** [/tmp/glibc-20170810-4202-d41b88/glibc-2.19/build/iconv/iconv_prog] Error 1
make[2]: *** Waiting for unfinished jobs....
/home/linuxbrew/.linuxbrew/opt/binutils/bin/ld: /tmp/glibc-20170810-4202-d41b88/glibc-2.19/build/csu/crt1.o: relocation R_X86_64_32S against symbol `__libc_csu_fini' can not be used when making a shared object; recompile with -fPIC
/home/linuxbrew/.linuxbrew/opt/binutils/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/6/crtbegin.o: relocation R_X86_64_32 against hidden symbol `__TMC_END__' can not be used when making a shared object
/home/linuxbrew/.linuxbrew/opt/binutils/bin/ld: final link failed: Nonrepresentable section on output
collect2: error: ld returned 1 exit status
../Rules:140: recipe for target '/tmp/glibc-20170810-4202-d41b88/glibc-2.19/build/iconv/iconvconfig' failed
make[2]: *** [/tmp/glibc-20170810-4202-d41b88/glibc-2.19/build/iconv/iconvconfig] Error 1
make[2]: Leaving directory '/tmp/glibc-20170810-4202-d41b88/glibc-2.19/iconv'
Makefile:213: recipe for target 'iconv/others' failed
make[1]: *** [iconv/others] Error 2
make[1]: Leaving directory '/tmp/glibc-20170810-4202-d41b88/glibc-2.19'
Makefile:9: recipe for target 'all' failed
make: *** [all] Error 2
```
See https://gist.github.com/anonymous/3e4d547cece375e813ac65b8bfa40646

I'm trying with an explicit `--cc=gcc-5` now to see if that fixes it.

(Also, I wouldn't read too much into my personal experience here; not only is Ubuntu 17.04 not the most common case and not somewhere a brewed `glibc` is all that useful, but also my configuration is fairly screwy from accumulated maintainer hacks).

**Comment by rwhogg** [comment]
Ok, that *sort of* worked. Installation was successful at that point, but the postinstall was not (and it didn't really say why...)

Furthermore, after doing this, a number of my brewed `coreutils` executables began segfaulting - probably because they had been built from source against my system Glibc 2.24.

Perhaps we should add a requirement to the `glibc` formula that your system Glibc is not newer than the Linuxbrew-packaged version? Otherwise, IIUC you'd have to recompile all your from-source formulae against the brewed `glibc` to regain binary compatibility.

**Comment by sjackman** [comment]
Thanks for pointing out this failure case. I hadn't really considered or tested the case of the brewed `glibc` being older than the host's. I'm in favour of adding a `Requirement` to prevent that.

**Comment by rwhogg** [comment]
I've opened #3791 for this case.

**Pull request closed without merging**